### ISSUE TITLE
Gradient for MatrixTriangularSolve 

### ISSUE THREAD

**Issue by alexggmatthews** [tensorflow/tensorflow#1875]
We have a gradient op in master for MatrixSolve . It seems like it shouldn't be too difficult to define a gradient for MatrixTriangularSolve in a similar way. Are there any imminent plans to do this? @rmlarsen @girving ?

I actually had a look at doing it myself but I was getting some strange numerical behaviour from the finite differences test. Possibly just something I am doing wrong. If we can work out what is going on with the test I would be happy to help.

**Comment by girving** [comment]
What numerical issues were you seeing, and what matrix were you differentiating around?  If the inverted matrix has bad condition number the gradients will be very unstable.

**Comment by alexggmatthews** [comment]
Hi Geoffrey,
Thanks for your quick response. I looked here for inspiration on the code:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py#L79

You need to account for the fact that the matrix will change from being upper triangular to lower triangular and vice versa when you take the adjoint but that's fine. 

I then played with adding the MatrixTriangularSolve op here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/linalg_grad_test.py#L133

and using np.tril to give it lower triangular random inputs.  I'm not sure I did this in a very tidy way. But in any case I was getting big differences between the finite differences and the backprop. Perhaps as you suggest random lower triangular matrices are very ill conditioned. 

I could submit as a PR with help needed if you like? I thought if you guys where doing this anyway that it might be more efficient not to tread on your toes.

**Comment by girving** [comment]
We're happy to look at a PR.  Everything you've written sounds reasonable, so the next steps are all details. :)

The main difference between lower triangular and full here is that the determinant of a triangular matrix is the product of the diagonal entries, so you have `n` entries where a zero makes the entire matrix singular regardless of the other entries.  I'd try picking the diagonal to have random values bounded well away from zero.

**Comment by rmlarsen** [comment]
As mentioned on the PR, I actually just submitted support for backprop for  {batch_}matrix_solve, {batch_}triangular_solve, and batch_matrix_determinant.
